### PR TITLE
New SOC formulation and self discharge

### PR DESCRIPTION
**Pull request by johburger**
Finalised the new SOC formulation that considers self discharge. Works in the PlanningTool, no testing of the platform tool has been done yet

### CODE REVIEW

File: projects/memapCore/src/main/java/memap/helper/milp/MILPProblemNoConnections.java [hunk 1]
```diff
@@ -295,7 +295,41 @@ public LpSolve createSOCBoundaries(LpSolve problem, BuildingMessage buildingMess
 			if (maxChargeCapacity >= sm.capacity) {
 				maxChargeCapacity = sm.capacity;
 			}
+			
+			// Hard coded standby losses
+			double standbyLosses = 0.2;
```
Comment (bytschkow) [inline projects/memapCore/src/main/java/memap/helper/milp/MILPProblemNoConnections.java:300]: shall we take this value from the storage messages and remove the hard coded stuff?

File: projects/memapCore/src/main/java/memap/messages/planning/StorageMessage.java [hunk 2]
```diff
@@ -22,6 +22,12 @@
 	public double efficiencyCharge;
 	public double efficiencyDischarge;
 	
+	public double storageLosses;
+	
+	public double alpha;
```
Comment (bytschkow) [inline projects/memapCore/src/main/java/memap/messages/planning/StorageMessage.java:27]: See also comment in the CVSStorage class:



The storageMessage is evisioned to contain only "device-parameters". 

Alpha and betas are calculated in the MPC part, hence, they should not be part of the storage message.

We sould even stick to the standard DIN EN 12897, which has the "Warmhalteverlust in W" label.

**Review by bytschkow — CHANGES_REQUESTED** [review]

File: projects/memapCore/src/main/java/memap/components/CSVStorage.java [hunk 3]
```diff
@@ -43,6 +43,10 @@ public void makeDecision() {
 		storageMessage.efficiencyCharge = effIN;
 		storageMessage.efficiencyDischarge = effOUT;
 		storageMessage.networkType = networkType;
+		storageMessage.storageLosses = storageLoss;
+		storageMessage.alpha = alpha;
```
[the remark this answers is not in the document]
Comment (johburger) [reply projects/memapCore/src/main/java/memap/components/CSVStorage.java:47]: Yes, alpha and beta values should be excluded here. I submitted the pull request before we discussed this. I will change this.

Regarding the storageLosses: We have to consider batteries as well, the DIN norm is only for thermal storages. For batteries the losses are typically given in %/Month at a certain temperature. I don't know what the best way to suit all possibilities is.

File: projects/memapCore/src/main/java/memap/helper/milp/MILPProblemNoConnections.java (hunk 1 above)
Comment (johburger) [reply projects/memapCore/src/main/java/memap/helper/milp/MILPProblemNoConnections.java:300]: As soon as the input field in the GUI is implemented, yes. Would be great if you could do this once the GUI is adjusted.

File: projects/memapCore/src/main/java/memap/components/CSVStorage.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (bytschkow) [reply projects/memapCore/src/main/java/memap/components/CSVStorage.java:47]: Yes, indeed, losses in %/month are a complete different magntitude than heat storage... for our use case we should assume 0 then. We should discuss that with Thomas and Jan, and see if they agree as well. From the GUI perspective - we can hide this property for batteries only.

[the remark this answers is not in the document]
Comment (JanAxelMayer) [reply projects/memapCore/src/main/java/memap/components/CSVStorage.java:47]: As I wrote in the Mail, the solution might be to introduce another String-field/datapoint for the units. Them we do a switch/case and perform the respective calculation depending on the given unit {Watt, Kwh/d, %/Month}.
However, due to time issues, I suggest to start with the special case in Watt, as discussed, and then later generalize it this way.

File: projects/memapCore/src/main/java/memap/helper/milp/MILPProblemNoConnections.java (hunk 1 above)
Comment (JanAxelMayer) [reply projects/memapCore/src/main/java/memap/helper/milp/MILPProblemNoConnections.java:300]: Denis is right. Both the GUI CSVStorage and the OPC UA ClientStorage send the storage message to the optimizer, so it should be read from the storage message. 
As long it is not implemented in the frontend, it should be rather hardcoded in the CSVStorage class and then fed into the storage message there.

File: projects/memapCore/src/main/java/memap/components/CSVStorage.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (bytschkow) [reply projects/memapCore/src/main/java/memap/components/CSVStorage.java:47]: Thanks for that comment, we can even implement an ENUM and a selection (similar as heat and electricity network type). I can do that.

[the remark this answers is not in the document]
Comment (thomaslicklederer) [reply projects/memapCore/src/main/java/memap/components/CSVStorage.java:47]: I agree with sticking to Watt for the standby losses as input value by user, as this is also according to DIN EN 12897. We should just have a look, if this is also the appropriate unit for batteries.

I would not do an enum or a case distinction, as for other input parameters we also force the user to insert the values in the correct units specified by us.

However, independent from the unit to be entered, we would need an additional reference SOC at which this standby losses are measured. For DIN EN 12897 it is SOC_ref = 0.75 (assuming that T_max is 80°C and T_min is 20°C). But should we just assume that the entered value (in Watt) always refers to the reference SOC of 0.75 or make an additional input? Perhaps for the sake of simplicity and due to time issues we should assume 0.75 as hardcoded standard reference SOC?

File: projects/memapCore/src/main/java/memap/messages/planning/StorageMessage.java (hunk 2 above)
Comment (thomaslicklederer) [reply projects/memapCore/src/main/java/memap/messages/planning/StorageMessage.java:27]: I agree with sticking to Watt for the standby losses as input value by user, as this is also according to DIN EN 12897. We should just have a look, if this is also the appropriate unit for batteries.

I would not do an enum or a case distinction, as for other input parameters we also force the user to insert the values in the correct units specified by us.

However, independent from the unit to be entered, we would need an additional reference SOC at which this standby losses are measured. For DIN EN 12897 it is SOC_ref = 0.75 (assuming that T_max is 80°C and T_min is 20°C). But should we just assume that the entered value (in Watt) always refers to the reference SOC of 0.75 or make an additional input? Perhaps for the sake of simplicity and due to time issues we should assume 0.75 as hardcoded standard reference SOC?

File: projects/memapCore/src/main/java/memap/components/prototypes/Storage.java [hunk 4]
```diff
@@ -34,7 +37,10 @@ public Storage(String name, double capacity, double stateOfCharge, double max_ch
 		this.max_discharging = max_discharging;
 		this.effIN = effIN;
 		this.effOUT = effOUT;
-		this.storageLoss = storageLoss;
+		// TODO remove hard coding.
+		// this.storageLoss = storageLoss;
+		
+		this.storageLoss = 0.1;
```
Comment (JanAxelMayer) [inline projects/memapCore/src/main/java/memap/components/prototypes/Storage.java:43]: This should be removed and changed back to this.storageLoss = storageLoss as Storage.java is an abstract class. Values are set in the inheriting class when instantiated.

File: projects/memapCore/src/main/java/memap/components/prototypes/Storage.java [hunk 5]
```diff
@@ -77,7 +83,11 @@ public void handleRequest() {
 			if (storageChargeRequest!= null && storageDischargeRequest!= null) {
```
Comment (JanAxelMayer) [inline projects/memapCore/src/main/java/memap/components/prototypes/Storage.java:83]: Same topic/question as above: Shouldn't this whole part better be shifted to CSVStrorage.java ?
The abstract class Storage.java defines the needed variables (and/or mehthods) but concrete numbers (and methods) are set when instantiated by the inheriting classes *CSVStorage.java* or *ClientStorages.java*.

As long as this input from the UI is missing, storage loss should be hardcoded in CSVStorage, not in the super-class.

Also renaming *linprogResult* to *optResult* may avoid confusion here.

**Review by JanAxelMayer — CHANGES_REQUESTED** [review]
The calculation of the future SOC (with the curr. Setpoint and the storage loss) should happen before giving the value into the storageMessage. 

This calculation should only be done in the classes CSVStorage.java (for the PlanningTool) and ClientStorage.java (for the opcua-based live optimization).

The optimizer should then use the updated SOC value from the storageMessage without further calculation for the next optimization-step (for setting the boundaries) without further calculation.

I'll add a commit with these changes.

Edit: Commit works with both live-version and Planning tool without errors, but correct storage behaviour was not tested yet.

**Review by bytschkow — APPROVED** [review]

**Comment by bytschkow** [comment]
The last version is running very well in the simulation. Therefore, the pull request can be accepted. However, some cleaning is still necessary. We can do it in another later pull request, though.

**Pull request merged**